server/pkcs11.go

```diff
@@ -0,0 +1,158 @@
+package server
+
+import (
+	"crypto"
+	"strconv"
+	"strings"
+	"net/url"
+	"regexp"
+
+	"github.com/thalesignite/crypto11"
+	"github.com/cloudflare/cfssl/log"
+)
+
+// PKCS11URI defines a PKCS#11 URI: https://tools.ietf.org/html/rfc7512#section-2.3
+type PKCS11URI struct { // pkcs11:path-attr[?query-attr]
+	// path-attr: (; delimited)
+	Token  string //        token <- CK_TOKEN_INFO
+	Manuf  string // manufacturer <- CK_TOKEN_INFO
+	Serial string //       serial <- CK_TOKEN_INTO
+	Model  string //        model <- CK_TOKEN_INFO
+
+	LibManuf string // library-manufacturer <- CK_INFO
+	LibDesc  string //  library-description <- CK_INFO
+	LibVer   string //      library-version <- CK_INFO
+
+	Object []byte // object <- CKA_LABEL
+	Type   string //   type <- CKA_CLASS (cert, data, private, public, or secret-key)
+	Id     []byte //     id <- CKA_ID
+
+	SlotManuf string // slot-manufacturer <- CK_SLOT_INFO
+	SlotDesc  string //  slot-description <- CK_SLOT_INFO
+	SlotId      uint //           slot-id <- CK_SLOT_ID
+
+	// query-attr: (& delimited)
+	PinSource string // pin-source
+	PinValue  string //  pin-value
+
+	ModuleName string // module-name
+	ModulePath string // module-path
+}
+
+func RFC7512Parser(uri string) PKCS11URI {
+	aChar := "[a-z-_]"
+	pChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=&]"
+	pAttr := aChar+"+="+pChar+"+"
+	pClause := "("+pAttr+";)*("+pAttr+")"
+	qChar := "[a-zA-Z0-9-_.~%:\\[\\]@!\\$'\\(\\)\\*\\+,=/\\?\\|]"
+	qAttr := aChar+"+="+qChar+"+"
+	qClause := "("+qAttr+"&)*("+qAttr+")"
+	r, _ := regexp.Compile("^pkcs11:"+pClause+"(\\?"+qClause+")?$")
+
+	if ! r.MatchString(uri) {
+		log.Error("PKCS#11 URI is malformed: " + uri)
```

Comment:
Instead of logging errors, the function signature should include an error return value, and you should return an error here. The calling code can decide about logging (think of this function like `strconv.Atoi`).